main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -497,10 +514,66 @@ object Inliner {
   /**
    * Convert a given tailCall expression `exp0` to a non tail call
    */
-  //TODO expand `convertTailCall` when more functions are being inlined
   private def convertTailCall(exp0: OccurrenceAst.Expression): OccurrenceAst.Expression = exp0 match {
+    case OccurrenceAst.Expression.Let(sym, exp1, exp2, occur, tpe, purity, loc) =>
+      val e2 = convertTailCall(exp2)
+      OccurrenceAst.Expression.Let(sym, exp1, e2, occur, tpe, purity, loc)
+
+    case OccurrenceAst.Expression.IfThenElse(exp1, exp2, exp3, tpe, purity, loc) =>
+      val e2 = convertTailCall(exp2)
+      val e3 = convertTailCall(exp3)
+      OccurrenceAst.Expression.IfThenElse(exp1, e2, e3, tpe, purity, loc)
+
+    case OccurrenceAst.Expression.Branch(e0, br0, tpe, purity, loc) =>
+      val br = br0 map {
+        case (sym, exp) => sym -> convertTailCall(exp)
+      }
+      OccurrenceAst.Expression.Branch(e0, br, tpe, purity, loc)
+
+    case OccurrenceAst.Expression.SelectChannel(rules, default, tpe, loc) =>
+      val rs = rules map {
+        case OccurrenceAst.SelectChannelRule(sym, chan, exp) => OccurrenceAst.SelectChannelRule(sym, chan, convertTailCall(exp))
+      }
+      val d = default.map(exp => convertTailCall(exp))
+      OccurrenceAst.Expression.SelectChannel(rs, d, tpe, loc)
+
     case OccurrenceAst.Expression.ApplyCloTail(exp, args, tpe, purity, loc) => OccurrenceAst.Expression.ApplyClo(exp, args, tpe, purity, loc)
+
     case OccurrenceAst.Expression.ApplyDefTail(sym, args, tpe, purity, loc) => OccurrenceAst.Expression.ApplyDef(sym, args, tpe, purity, loc)
+
+    case OccurrenceAst.Expression.ApplySelfTail(sym, _, actuals, tpe, purity, loc) => OccurrenceAst.Expression.ApplyDef(sym, actuals, tpe, purity, loc)
+
+    case _ => exp0
+  }
+
+  /**
+   * Convert a given ApplySelfTail expression `exp0` to a ApplyDefTail
+   */
+  private def convertSelfTailCall(exp0: OccurrenceAst.Expression): OccurrenceAst.Expression = exp0 match {
```

Comment:
I am a bit spooked by this. 

Do you understand what this does? It seems to me this corresponds to unrolling a recursive loop once. Is that what you want? Is that "good"?

Reply:
It was used when inlining a self recursive function into the body of another function. 
```
def foo(x: Int32): Int32 = 
    if (x > 5)
        1
    else
        foo(x-1)

def bar(x: Int32): Int32 = foo(x)
```

If we wanted to inline foo into bar we would have to convert the `SelfTailCall` in foo into a `ApplyDefTail`

But this is not something we inline right now